### ISSUE TITLE
Incorrect CS8762 when using `[NotNullWhen(true)]` and static inner functions

### ISSUE THREAD

**Issue by yufeih** [dotnet/roslyn#42523]
The following code reports `lib\HtmlUtility.cs(249,17): error CS8762: Parameter 'title' may not have a null value when exiting with 'true'. [C:\Code\docfx\src\docfx\docfx.csproj]`.

If I move out the inner function `IsInvisibleNode`, the warning goes away.

```csharp

        public static bool TryExtractTitle(HtmlNode node, [NotNullWhen(true)] out string? title, [NotNullWhen(true)] out string? rawTitle)
        {
            var existVisibleNode = false;

            title = null;
            rawTitle = string.Empty;
            foreach (var child in node.ChildNodes)
            {
                if (!IsInvisibleNode(child))
                {
                    if (child.NodeType == HtmlNodeType.Element && (child.Name == "h1" || child.Name == "h2" || child.Name == "h3"))
                    {
                        title = HttpUtility.HtmlDecode(child.InnerText ?? "");

                        // NOTE: for backward compatibility during migration phase, the logic of title and raw title is different...
                        if (!existVisibleNode)
                        {
                            rawTitle = child.OuterHtml;
                            child.Remove();
                        }

                        return true;
                    }

                    existVisibleNode = true;
                }
            }

            return false;

            static bool IsInvisibleNode(HtmlNode n)
            {
                return n.NodeType == HtmlNodeType.Comment ||
                    (n.NodeType == HtmlNodeType.Text && string.IsNullOrWhiteSpace(n.OuterHtml));
            }
        }
```

**Comment by jcouv** [comment]
Apologies. This is a known bug in 16.6p1 which was fixed in 16.6p2, reported in issue https://github.com/dotnet/roslyn/issues/42199 and fixed in PR https://github.com/dotnet/roslyn/pull/41675

The problem is that we analyze `return` statements in the method, to enforce the state of parameters marked with `[NotNullWhen(true)]`, but spuriously check the `return` statements inside local functions...

I'll go ahead and close as duplicate.

**Comment by yufeih** [comment]
No worries, I just upgrade to `3.1.200` today and noticed this. Glad this is already fixed 👍
